easypermissions/src/main/java/pub/devrel/easypermissions/EasyPermissions.java

```diff
@@ -85,205 +80,220 @@ public static boolean hasPermissions(@NonNull Context context, @NonNull String..
      *
-     * @param object
-     *         Activity or Fragment requesting permissions. Should implement
-     *         {@link android.support.v4.app.ActivityCompat.OnRequestPermissionsResultCallback}
-     *         or {@code android.support.v13.app.FragmentCompat.OnRequestPermissionsResultCallback}
-     * @param rationale
-     *         a message explaining why the application needs this set of permissions, will be displayed if the user rejects the request the first
-     *         time.
-     * @param positiveButton
-     *         custom text for positive button
-     * @param negativeButton
-     *         custom text for negative button
-     * @param requestCode
-     *         request code to track this request, must be < 256.
-     * @param perms
-     *         a set of permissions to be requested.
+     * @param activity       Activity requesting permissions. Should implement {@link
+     *                       ActivityCompat.OnRequestPermissionsResultCallback} {@code
+     *                       android.support.v13.app.FragmentCompat.OnRequestPermissionsResultCallback}
+     * @param rationale      a message explaining why the application needs this set of permissions,
+     *                       will be displayed if the user rejects the request the first time.
+     * @param positiveButton custom text for positive button
+     * @param negativeButton custom text for negative button
+     * @param requestCode    request code to track this request, must be < 256.
+     * @param perms          a set of permissions to be requested.
+     * @see android.Manifest.permission
      */
-    @SuppressLint("NewApi")
-    public static void requestPermissions(@NonNull final Object object, @NonNull String rationale,
-                                          @StringRes int positiveButton, @StringRes int negativeButton,
-                                          final int requestCode, @NonNull final String... perms) {
-
-        checkCallingObjectSuitability(object);
-
-        // Determine if rationale should be shown (generally when the user has previously
-        // denied the request);
-        boolean shouldShowRationale = false;
-        for (String perm : perms) {
-            shouldShowRationale =
-                    shouldShowRationale || shouldShowRequestPermissionRationale(object, perm);
+    @SuppressWarnings("NewApi")
+    public static void requestPermissions(@NonNull Activity activity,
+                                          @NonNull String rationale,
+                                          @StringRes int positiveButton,
+                                          @StringRes int negativeButton,
+                                          int requestCode,
+                                          @NonNull String... perms) {
+        if (hasPermissions(activity, perms)) {
+            notifyAlreadyHasPermissions(activity, requestCode, perms);
+            return;
         }
```

Comment:
Ok so I think the use of `notifyAlreadyHasPermissions` could be a breaking change for some developers.

Let's say you have this code

```java
private foid foo() {
  if (!EasyPermissions.hasPermissions(this, "BAR")) {
    EasyPermissions.requestPermissions(this, "BAR");
    return;
  }

   // Do some important stuff
}

@Override
public void onPermissionsGranted(List<String> perms) {
  if (perms.contains("BAR")) {
    foo();
  }
}
```

The problem is that by notifying them of already having the permission, this could result in `foo()` being called twice.  For some operations, this could be very bad (anything that's not idempotent).

So I think instead of notifying listeners when they already have permissions, we may have to just Log a warning and then drop the whole thing in order to preserve current behavior.